[SYCL] Unify usm memcpy/memset behavior for corner cases

v-klochkov opened the pull request:
This patch handles queue::memcpy, queue::memset, handler::memcpy,
handler::memset.

Currently, queue::memcpy(nullptr, s, num_bytes) gives SegFault for HOST
and exception throw for CPU/GPU. With this patch an exception is thrown
ealier for all devices. Same for memset.

The call of queue::memcpy(nullptr, s, 0) is skipped before such call reaches device.
That is in sync with std::memcpy(nullptr, s, 0), which is deleted by clang/gcc
even with -O0 when num_of_bytes is statically known to be 0.

Signed-off-by: Vyacheslav N Klochkov <vyacheslav.n.klochkov@intel.com>

keryell commented:
All this deep subtle useful PR description should be as design comments in the code.
This is open-source code to be read & *understood* by many people, not write-only code. :-)

jbrodman commented:
LGTM, but please add inline comments like @keryell suggested.

v-klochkov commented on sycl/test/usm/queue_wait.cpp:
```diff
@@ -38,11 +37,28 @@ int main() {
+  try {
+    Q.memset(nullptr, 42, Size);
+    Q.wait_and_throw();
+  } catch (runtime_error e) {
+  }
```
Ok, I added an assert to the try section. Thank you.

v-klochkov commented:
Thank you for the code-review! 
In the 3rd commit I updated the LIT test (added assert to check that exception is thrown) and added comment sections requested by @keryell and @jbrodman

v-klochkov commented:
> Qưe

I don't know what this 'change request' means, but it helped me to find a misprint copy-pasted 4 times: "behavoir" -> "behavior".

jbrodman commented:
> > Qưe
> 
> I don't know what this 'change request' means, but it helped me to find a misprint copy-pasted 4 times: "behavoir" -> "behavior".

Ha.  That account seems to be just leaving spam comments on various PRs of open source projects...

sergey-semenov commented on sycl/include/CL/sycl/handler.hpp:
```diff
@@ -1732,13 +1732,19 @@ class __SYCL_EXPORT handler {
 
   /// Copies data from one memory region to another, both pointed by
   /// USM pointers.
+  /// No operations is done if \param Count is zero. An exception is thrown
```
Minor: typo in all of the 4 comments, "no operations is" -> "no operation is"/"no operations are"

bader replied:
@v-klochkov, please, do not forget to fix the grammar in the comments.
